### PR TITLE
Some fixes for common bugs in HttpClient

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/common/http/test/response_spec.ts [hunk 1]
```diff
@@ -40,6 +40,11 @@ export function main() {
         expect(resp.ok).toBeTruthy();
         expect(resp.url).toBeNull();
       });
+      it('accepts a falsy body', () => {
+        const resp = new HttpResponse({body: false});
```
Comment (gkalpak) [inline packages/common/http/test/response_spec.ts]: Unused?

File: packages/common/http/src/xhr.ts [hunk 2]
```diff
@@ -180,24 +180,20 @@ export class HttpXhrBackend implements HttpBackend {
...
-            body = { error, text: body } as HttpJsonParseError;
-          }
-        } else if (!ok && req.responseType === 'json' && typeof body === 'string') {
-          try {
-            // Attempt to parse the body as JSON.
-            body = JSON.parse(body);
-          } catch (error) {
-            // Cannot be certain that the body was meant to be parsed as JSON.
-            // Leave the body as a string.
+            // If this was an error request to begin with, leave it as a string, it probably
```
Comment (gkalpak) [inline packages/common/http/src/xhr.ts]: I wonder if it is worth preserving the `XSSI_PREFIX` in this case 🤔

**Review by gkalpak — COMMENTED** [review]
The 4th commit also closes PR #19917 btw.

File: packages/common/http/test/response_spec.ts (hunk 1 above)
Comment (alxhub) [reply packages/common/http/test/response_spec.ts]: Removed.

File: packages/common/http/src/xhr.ts (hunk 2 above)
Comment (alxhub) [reply packages/common/http/src/xhr.ts]: Yes, done in a new commit.

**Review by gkalpak — APPROVED** [review]

**Review by gkalpak — COMMENTED** [review]
The 4th commit has incorrect scope (`service-worker` instead of `common`).

**Comment by quanterion** [comment]
When is it going to be merged? Without those fixes it's impossible to use most HTTP DELETE requests that returns empty body

**Comment by Maximaximum** [comment]
Hopefully this PR will get released as soon as possible, because the empty body parse issue is a real show stopper.

**Comment by Taha-Di-Nero** [comment]
When this fix will be released ???
I spent to much time to migrate to the new HttpClient to downgrade.................

**Comment by valkoun** [comment]
FYI, in the meantime you have 3 options to workaround:

1) Make sure responses always return a response body with 200 responses.
2) Make each response with no response body return a 204 No Content code.
3) Set `responseType: 'text'` for each http method call:
```
return this.http.post('api/groups', group, {
      responseType: 'text',
});
```

**Comment by Maximaximum** [comment]
@valkoun yeah, thanks, but it's a whole lot of stupid error-prone work, so I hope this fix will get into release ASAP

**Comment by Taha-Di-Nero** [comment]
Meanwhile this will be merged,  I fixed the problem on the server side returning http status 204 for all void service method and so no more parsing error.

**Comment by realshadow** [comment]
Will this also get merged back to 4.4.6 or 7?

**Comment by gigioSouza** [comment]
When will it be released?

**Comment by vinayakpatil** [comment]
@alxhub Will this be merged anytime soon?

**Comment by alxhub** [comment]
Yes, hopefully next week.

**Review by vikerman — APPROVED** [review]

**Comment by hspier** [comment]
Any chance we can get a 4.4.7 release with this fix soon?

**Pull request closed without merging**